includes/class-wc-payments-webhook-processing-service.php

```diff
@@ -783,4 +787,71 @@ private function get_failure_message_from_error( $error ):string {
 		// translators: %s Stripe error message.
 		return sprintf( __( 'With the following message: <code>%s</code>', 'woocommerce-payments' ), $message );
 	}
+
+	/**
+	 * Process webhook refund event.
+	 *
+	 * @param array $event_body The event that triggered the webhook.
+	 *
+	 * @throws Invalid_Webhook_Data_Exception           Required parameters not found.
+	 * @throws Invalid_Webhook_Data_Exception           When the refund amount is not valid.
+	 * @throws Order_Not_Found_Exception                When unable to resolve charge ID to order.
+	 */
+	private function process_webhook_refund( array $event_body ): void {
+		$event_data   = $this->read_webhook_property( $event_body, 'data' );
+		$event_object = $this->read_webhook_property( $event_data, 'object' );
+
+		$is_refunded_event = isset( $event_body['type'] ) && 'charge.refunded' === $event_body['type'];
+		$status            = $this->read_webhook_property( $event_object, 'status' );
+		if ( 'succeeded' !== $status || ! $is_refunded_event ) {
+			return;
+		}
+
+		// Fetch the details of the refund so that we can find the associated order and write a note.
+		$charge_id                     = $this->read_webhook_property( $event_object, 'id' );
+		$refund                        = $this->read_webhook_property( $event_object, 'refunds' )['data'][0]; // Most recent refund.
+		$refund_id                     = $refund['id'] ?? '';
+		$refund_reason                 = $refund['reason'] ?? '';
+		$refund_balance_transaction_id = $refund['balance_transaction'] ?? '';
+		$charge_amount                 = $this->read_webhook_property( $event_object, 'amount' );
+		$currency                      = $this->read_webhook_property( $event_object, 'currency' );
+		$refunded_amount               = WC_Payments_Utils::interpret_stripe_amount( $refund['amount'], $currency );
+		$is_partial_refund             = $refund['amount'] < $charge_amount;
+
+		// Look up the order related to this charge.
+		$order = $this->wcpay_db->order_from_charge_id( $charge_id );
+		if ( ! $order ) {
+			throw new Order_Not_Found_Exception(
+				sprintf(
+				/* translators: %1: charge ID */
+					__( 'Could not find order via charge ID: %1$s', 'woocommerce-payments' ),
+					$charge_id
+				),
+				'order_not_found'
+			);
+		}
+		// This is to avoid processing the same refund multiple times in case of a webhook retry.
```

Comment:
More precisely: 
```suggestion
		// Only care about refunds that are triggered externally, i.e. outside WP Admin.
```

IMO, we would also change this method name to `process_webhook_refund_trigged_externally`. Though it's pretty long but it conveys the full purpose of this PR as well as this method.

Reply:
Thanks for the suggestion, I agree that it's more clear this way. I also expanded the comments to point to the place where WP Admin refunds are handled. See [9224dc1](https://github.com/Automattic/woocommerce-payments/pull/8200/commits/9224dc1f71e12ad1e908f9545c7b5e7f58459984).